File: frontend/src/lib/types/branch.ts
export interface BranchCreate {
  name: string;
  rp_folder: string;
  description?: string;
  branch_from?: string;
  branch_point_exchange?: number;
}
export interface BranchInfo {
  name: string;
  rp_folder: string;
  created_from: string | null;
  branch_point_session: string | null;
  branch_point_exchange: number | null;
  description: string | null;
  is_active: boolean;
  is_archived: boolean;
  created_at: string | null;
  exchange_count: number;
}
export interface BranchListResponse {
  active_branch: string | null;
  branches: BranchInfo[];
}
export interface BranchSwitchRequest {
  name: string;
}
export interface BranchSwitchResponse {
  active_branch: string;
  previous_branch: string | null;
}
export interface CheckpointCreate {
  name: string;
  description?: string;
}
export interface CheckpointInfo {
  name: string;
  branch: string;
  exchange_number: number;
  description: string | null;
  created_at: string;
}
export interface CheckpointRestoreRequest {
  checkpoint_name: string;
}
export interface CheckpointRestoreResponse {
  restored_from: string;
  exchange_number: number;
  new_branch: string | null;
  rewound_count: number | null;
}

File: rp_engine/models/branch.py
"""Pydantic models for branch management."""

from __future__ import annotations

from pydantic import BaseModel


class BranchCreate(BaseModel):
    name: str
    rp_folder: str
    description: str | None = None
    branch_from: str | None = None
    branch_point_exchange: int | None = None


class BranchInfo(BaseModel):
    name: str
    rp_folder: str
    created_from: str | None = None
    branch_point_session: str | None = None
    branch_point_exchange: int | None = None
    description: str | None = None
    is_active: bool = False
    is_archived: bool = False
    created_at: str | None = None
    exchange_count: int = 0


class BranchArchiveRequest(BaseModel):
    archived: bool = True


class BranchListResponse(BaseModel):
    active_branch: str | None = None
    branches: list[BranchInfo]


class BranchSwitchRequest(BaseModel):
    name: str


class BranchSwitchResponse(BaseModel):
    active_branch: str
    previous_branch: str | None = None


class CheckpointCreate(BaseModel):
    name: str
    description: str | None = None


class CheckpointInfo(BaseModel):
    name: str
    branch: str
    exchange_number: int
    description: str | None = None
    created_at: str


class CheckpointRestoreRequest(BaseModel):
    checkpoint_name: str


class CheckpointRestoreResponse(BaseModel):
    restored_from: str
    exchange_number: int
    new_branch: str | None = None
    rewound_count: int | None = None  # deprecated, kept for backward compat
